x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/api/perform_rule_upgrade/diffable_rule_fields_mappings.ts

```diff
@@ -0,0 +1,211 @@
+
+  const transformedValue = transformDiffableFieldValues(field, diffableField);
+  if (transformedValue.type === 'TRANSFORMED_FIELD') {
+    return transformedValue.value;
+  }
+
+  // From the ThreeWayDiff, get the specific field that maps to the diffable rule field
+  // Otherwise, the diffableField itself already matches the rule field, so retrieve that value.
+  const mappedField = get(diffableField, diffableFieldMapping, diffableField);
+
+  return mappedField;
+};
+
+interface GetGroupNameForFieldParams {
+  ruleType: DiffableRuleTypes;
+  fieldName: string;
+}
+/**
+ * Maps a PrebuiltRuleAsset schema field name to its corresponding DiffableRule group.
+ *
+ * Determines which group in the DiffableRule schema a given field belongs to. Handles special
+ * cases for query-related fields based on the rule type.
+ *
+ * @param {string} fieldName - The field name from the PrebuiltRuleAsset schema.
+ * @param {string} ruleType - The type of the rule being processed.
+ *
+ * @example
+ * getGroupNameForField('index', 'query')
+ * // Returns: 'data_source'
+ *
+ * @example
+ * getGroupNameForField('query', 'eql')
+ * // Returns: 'eql_query'
+ *
+ */
+export function getGroupNameForField({
+  ruleType,
+  fieldName,
+}: GetGroupNameForFieldParams): keyof AllFieldsDiff {
+  const groupMapping: Record<string, keyof AllFieldsDiff> = {
+    building_block_type: 'building_block',
+    saved_id: 'kql_query',
+    threat_query: 'threat_query',
+    threat_language: 'threat_query',
+    threat_filters: 'threat_query',
+    index: 'data_source',
+    data_view_id: 'data_source',
+    rule_name_override: 'rule_name_override',
+    interval: 'rule_schedule',
+    from: 'rule_schedule',
+    to: 'rule_schedule',
+    timeline_id: 'timeline_template',
+    timeline_title: 'timeline_template',
+    timestamp_override: 'timestamp_override',
+    timestamp_override_fallback_disabled: 'timestamp_override',
+  };
+
+  // Handle query, filters and language fields based on rule type
+  if (fieldName === 'query' || fieldName === 'language' || fieldName === 'filters') {
+    switch (ruleType) {
+      case 'query':
+      case 'saved_query':
+        return 'kql_query' as const;
+      case 'eql':
+        return 'eql_query';
+      case 'esql':
+        return 'esql_query';
+      default:
+        return 'kql_query';
+    }
+  }
+
+  return groupMapping[fieldName] || fieldName;
+}
+
+/**
+ * Maps a PrebuiltRuleAsset schema field name to its corresponding property
+ * name within a DiffableRule group.
+ *
+ * @param {string} fieldName - The field name from the PrebuiltRuleAsset schema.
+ * @returns {string} The corresponding property name in the DiffableRule group.
+ *
+ * @example
+ * getDiffableFieldMapping('index')
+ * // Returns: 'index_patterns'
+ *
+ * @example
+ * getDiffableFieldMapping('from')
+ * // Returns: 'lookback'
+ *
+ */
+export function getDiffableFieldMapping(fieldName: string): string {
```

Comment:
Naming here is a little bit more complex than above, where we renamed to `mapRuleFieldToDiffableRuleField`, because we are not mapping to `Diffable Rule Fields`, but to each of the properties within each of the `diffable rule fields`, which do not match 100% to neither the rule schema nor the diffable rule schema.

For example, `lookback` is just a subfield within the `rule_schedule` diffable rule field, but doesn't exist in the rule schema. Same for `index_pattern` whose "corresponding" field in the rule schema is actually `index`.

For now I'll rename it to `mapRuleFieldToDiffableRuleSubfield`, let me know what you think.